[DROOLS-7201] Migration guide : non executable model to executable model (#4745)
kiegroup/drools: drools-docs/src/modules/ROOT/pages/migration-guide/_nonexec-model-to-exec-model.adoc

[id='nonexec-model-to-exec-model_{context}']

> == Migration from non executable model to executable model

Reviewer: ```suggestion
== Migration from non-executable model to executable model
```

> `drools-engine-classic` and `drools-mvel` execute rules with `non executable model`. `drools-engine` executes rules with `executable model`. Refer to xref:KIE/index.adoc#executable-model-con_packaging-deploying[Executable rule models] for executable model details.

Reviewer 2: ```suggestion
The `drools-engine-classic` and `drools-mvel` execute rules with `non executable model`. `drools-engine` executes rules with `executable model`. Refer to xref:KIE/index.adoc#executable-model-con_packaging-deploying[Executable rule models] for executable model details.
```

very minor just to avoiding starting with typescript font

> The `drools-engine-classic` and `drools-mvel` execute rules with `non executable model`. `drools-engine` executes rules with `executable model`. Refer to xref:KIE/index.adoc#executable-model-con_packaging-deploying[Executable rule models] for executable model details.

Reviewer: ```suggestion
The `drools-engine-classic` and `drools-mvel` execute rules with `non executable model`. `drools-engine` executes rules with `executable model`. For more details about the executable model, see xref:KIE/index.adoc#executable-model-con_packaging-deploying[Executable rule models].
```

> As we mentioned, `drools-engine-classic` and `drools-mvel` are deprecated, so we recommend you to migrate from `non executable model` to `executable model` by changing dependencies.

Reviewer: ```suggestion
As mentioned, `drools-engine-classic` and `drools-mvel` are deprecated, it is recommended to migrate from a `non-executable model` to an `executable model` by changing dependencies.
```

> Basically, rule executions results are the same with both `non executable model` and `executable model`. If you find any issues, please report it to {PRODUCT} community. However, there are a few known differences.

Reviewer 2: ```suggestion
In general, from an external perspective, rule evaluation and rule consequence executions results can be expected to be the same using either `non executable model` and `executable model`. You may however encounter a few edge cases during compile time or runtime; if you find any issues, please do report it to {PRODUCT} community. We have already identified some small known differences, reported below.
```

=== Invalid cast

> When you specify `dialect "mvel"` in a rule, `non executable model` is tolerant of invalid type cast. For example,

Reviewer 2: ```suggestion
When you specify `dialect "mvel"` in a rule, `non executable model` is tolerant of invalid type cast, because of type coercion behaviour in MVEL. For example,
```

> In general, from an external perspective, rule evaluation and rule consequence executions results can be expected to be the same using either `non executable model` and `executable model`. You may however encounter a few edge cases during compile time or runtime; if you find any issues, please do report it to {PRODUCT} community. We have already identified some small known differences, reported below.

Reviewer: ```suggestion
In general, from an external perspective, rule evaluation and rule consequence execution results can be expected to be the same using either the `non-executable` model or the `executable model`. You may however encounter a few edge cases during compile time or runtime. If you find any issues, please report them to the {PRODUCT} community. We have already identified some small known differences, reported below.
```

> When you specify `dialect "mvel"` in a rule, `non executable model` is tolerant of invalid type cast, because of type coercion behaviour in MVEL. For example,

Reviewer: ```suggestion
When you specify `dialect "mvel"` in a rule, the `non-executable model` is tolerant of invalid type cast, because of type coercion behavior in MVEL. For example,
```

[source]
----
global MyUtils myUtils;

rule Rule1
dialect "mvel"
when
  $p : Person()
then
  myUtils.doWork((String) $p.age )));
end
----
This rule can be built with  `non executable model` even if `$p.age` is `int` that is not acceptable in Java syntax.

With `executable model`, the rule fails with a build error.
----
Cannot cast from int to String
----

> In this case, the rule needs to be fixed. For example,

Reviewer 2: ```suggestion
In this case, the rule can be easily fixed by using valid Java syntax. For example,
```

> This rule can be built with  `non executable model` even if `$p.age` is `int` that is not acceptable in Java syntax.

Reviewer: ```suggestion
This rule can be built with  `non-executable model` even if `$p.age` is `int` which is not acceptable in Java syntax.
```
